pkg/scheduler/metrics/metric_recorder.go

```diff
@@ -77,3 +79,79 @@ func (r *PendingPodsRecorder) Dec() {
 func (r *PendingPodsRecorder) Clear() {
 	r.recorder.Set(float64(0))
 }
+
+// metric is the data structure passed in the buffer channel between the main framework thread
+// and the metricsRecorder goroutine.
+type metric struct {
+	metric      *k8smetrics.HistogramVec
+	labelValues []string
+	value       float64
+}
+
+// MetricAsyncRecorder records metric in a separate goroutine to avoid overhead in the critical path.
+type MetricAsyncRecorder struct {
+	// bufferCh is a channel that serves as a metrics buffer before the metricsRecorder goroutine reports it.
+	bufferCh chan *metric
+	// if bufferSize is reached, incoming metrics will be discarded.
+	bufferSize int
+	// how often the recorder runs to flush the metrics.
+	interval time.Duration
+
+	// stopCh is used to stop the goroutine which periodically flushes metrics.
+	stopCh <-chan struct{}
+	// IsStoppedCh indicates whether the goroutine is stopped. It's used in tests only to make sure
+	// the metric flushing goroutine is stopped so that tests can collect metrics for verification.
+	IsStoppedCh chan struct{}
+}
+
+func NewMetricsAsyncRecorder(bufferSize int, interval time.Duration, stopCh <-chan struct{}) *MetricAsyncRecorder {
+	recorder := &MetricAsyncRecorder{
+		bufferCh:    make(chan *metric, bufferSize),
+		bufferSize:  bufferSize,
+		interval:    interval,
+		stopCh:      stopCh,
+		IsStoppedCh: make(chan struct{}),
+	}
+	go recorder.run()
+	return recorder
+}
+
+// ObservePluginDurationAsync observes the plugin_execution_duration_seconds metric.
+// The metric will be flushed to Prometheus asynchronously.
+func (r *MetricAsyncRecorder) ObservePluginDurationAsync(extensionPoint, pluginName, status string, value float64) {
```

Comment:
is it intentional to change the type of status from Status to string?

Reply:
Yes. We cannot use `k8s.io/kubernetes/pkg/scheduler/framework` here due to the import cycle.

```
# When we import k8s.io/kubernetes/pkg/scheduler/framework here,
# and build the scheduler.

package k8s.io/kubernetes/pkg/scheduler/metrics
	imports k8s.io/kubernetes/pkg/scheduler/framework
	imports k8s.io/kubernetes/pkg/scheduler/framework/parallelize
	imports k8s.io/kubernetes/pkg/scheduler/metrics: import cycle not allowed
```